### PR TITLE
allow setting remappings on lsp server init

### PR DESCRIPTION
**Pull request by chrisfarms**
### What

enables client's of the lsp server to configure remappings during the initialisation command.

we already allow configuring import-paths through the initialisation, so hopefully this is not controversial.

the format expected to be passed in from the initialise json is:

```
{ 
  "remappings": [
      {context: "", prefix: "foo", target: "bar"},
      ...
  ]
}
```

### Why

Because without being able to set remappings, the lsp-server throws errors for any project that requires them.

### Related

* https://github.com/ethereum/solidity/issues/12603
* https://github.com/ethereum/solidity/pull/12994

### CODE REVIEW

**Comment by christianparpart** [comment]
Many thanks for your contribution @chrisfarms. We will take a close look at your work. :)

File: libsolidity/lsp/FileRepository.h
Comment (christianparpart) [inline libsolidity/lsp/FileRepository.h]:
```suggestion
	/// Remappings of imports.
```

File: libsolidity/lsp/LanguageServer.cpp
Comment (christianparpart) [inline libsolidity/lsp/LanguageServer.cpp]: includes are grouped by category and per category alphabetically sorted. I think this line should go 2 lines up, above ReadFile, if I'm not mistaken.

File: libsolidity/lsp/LanguageServer.cpp
Comment (christianparpart) [inline libsolidity/lsp/LanguageServer.cpp]:
```suggestion
					remappings.emplace_back(ImportRemapper::Remapping{
						jsonPath.get("context", "").asString(),
						jsonPath.get("prefix", "").asString(),
						jsonPath.get("target", "").asString(),
					});
```

`frontend` namespace was imported on top of the file already and  constructing the parameter directly as part of the emplace-call will not make you accidentally forget to "move" the temporary.

File: libsolidity/lsp/LanguageServer.cpp
Comment (christianparpart) [inline libsolidity/lsp/LanguageServer.cpp]: I think that's subjective, but maybe it's worth not just mentioning "array of objects" but rather say what that object should be like, especially since it's only three members that it must contain. But then it should be most useful to the reader.

**Review by christianparpart — CHANGES_REQUESTED** [review]
I am sure you did your own interactive tests to make sure it all works as expected. 
But maybe we also have that automatically tested as part of the still young LSP CI tests.

For that you might want to look at `test/lsp.py` and its companion test dir `test/libsolidity/lsp/`. In case of concerns, @Marenz can be of great help with respect to that. :)

**Comment by chrisfarms** [comment]
Thanks, addressed your review comments.

Will take a look at the lsp.py tests shortly (had not spotted those tests) 👍

**Comment by chrisfarms** [comment]
added a very basic case to the python lsp tests

File: test/lsp.py
Comment (Marenz) [inline test/lsp.py]: It looks to me like you need the content of this file only in this function and not for any of the "in file defined" test mechanics
Given that it serves no purpose other than in this function here, it would be nicer to just embed file test content directly like it is done in the function `test_textDocument_opening_two_new_files_edit_and_close`

You can still use the `otherlib.sol` for the import of course.

Comment (chrisfarms) [reply test/lsp.py]: 👍

**Comment by Marenz** [comment]
I rebased the PR to the latest `develop` branch to fix the pylint CI failure

**Review by Marenz — DISMISSED** [review]

File: libsolidity/lsp/LanguageServer.cpp
Comment (leonardoalt) [inline libsolidity/lsp/LanguageServer.cpp]: Nitpick: I think our style guide says no curly braces for single statement if/else blocks.

File: libsolidity/lsp/LanguageServer.cpp
Comment (leonardoalt) [inline libsolidity/lsp/LanguageServer.cpp]: This never goes negative so could/should be size_t/unsigned.

File: libsolidity/lsp/LanguageServer.cpp
Comment (leonardoalt) [inline libsolidity/lsp/LanguageServer.cpp]: This number is not used to `break` the loop at any point and only used as a `bool` later, so I think it should be a bool instead.

**Review by leonardoalt — CHANGES_REQUESTED** [review]

File: libsolidity/lsp/LanguageServer.cpp
Comment (leonardoalt) [inline libsolidity/lsp/LanguageServer.cpp]:
```suggestion
			m_fileRepository.setRemappings(std::move(remappings));
```

**Comment by christianparpart** [comment]
@chrisfarms I have been taking over this PR to finish the review, simply applying @leonardoalt's comments & adding a changelog item. I hope you don't mind.

**Review by Marenz — DISMISSED** [review]

**Comment by chrisfarms** [comment]
> @chrisfarms I have been taking over this PR to finish the review, simply applying @leonardoalt's comments & adding a changelog item. I hope you don't mind.

🙏 don't mind at all, thanks!

File: Changelog.md
Comment (leonardoalt) [inline Changelog.md]:
```suggestion
 * Language Server: Add configuration option to apply custom remappings via ``remappings`` in the LSP settings object.
```
Nitpick, but the gerund is not the standard in the changelog

File: libsolidity/lsp/LanguageServer.cpp
Comment (leonardoalt) [inline libsolidity/lsp/LanguageServer.cpp:236]: What if these are not strings? Shouldn't this also imply `typeFailures = true`?

Comment (leonardoalt) [reply libsolidity/lsp/LanguageServer.cpp:236]: It would also be nice to have tests for these.

**Review by leonardoalt — CHANGES_REQUESTED** [review]
Need some clarification

**Comment by christianparpart** [comment]
Adding @cameel for the final review.

File: libsolidity/lsp/LanguageServer.cpp
Comment (cameel) [inline libsolidity/lsp/LanguageServer.cpp:232]: Should we report an error on unexpected keys?

File: test/lsp.py
Comment (cameel) [inline test/lsp.py:1544]:
```suggestion
    def test_remapping_wrong_parameters(self, solc: JsonRpcProcess) -> None:
```
Also, something like `test_remapping_wrong_parameter_types` would be more precise.

File: libsolidity/lsp/LanguageServer.cpp
Comment (cameel) [inline libsolidity/lsp/LanguageServer.cpp:232]: This is missing a validation that `target` is not empty. solc won't accept remappings with an empty target on the CLI.

File: test/lsp.py
Comment (cameel) [inline test/lsp.py:1605]: I think that one of these `"no diagnostics"` messages does not match the number :)

File: test/lsp.py
Comment (cameel) [inline test/lsp.py:1583]: It would be best to set this to something that would match the remapping if used in an import to show that these URIs are not remapped.

**Review by cameel — COMMENTED** [review]
Only found minor stuff. Overall the feature seems sound.

**Comment by cameel** [comment]
Does this fix #12603? If so, please add `#Fixes #12603` in the PR description.

Also needs rebase on latest `develop` and approvals from @christianparpart and @leonardoalt.

**Comment by nikola-matic** [comment]
Rebased and resolved conflicts, should be in a more workable state now.

File: libsolidity/lsp/LanguageServer.cpp
Comment (nikola-matic) [inline libsolidity/lsp/LanguageServer.cpp:237]:
```suggestion
					remappings.emplace_back(
						jsonPath.get("context", "").asString(),
						jsonPath.get("prefix", "").asString(),
						jsonPath.get("target", "").asString(),
					);
```

File: libsolidity/lsp/LanguageServer.cpp
Comment (nikola-matic) [inline libsolidity/lsp/LanguageServer.cpp:229]: New type checking helpers were added and recently merged in [this pull request](https://github.com/ethereum/solidity/pull/13579/files#diff-56f32f3c6823f15db6b0437642c7639a500cf3af3d7a86e16db2073dffca60ceR79), so these should be changed to use `isOfType<T>` or `isOfTypeIfExists<T>`, whichever one is more appropriate.

File: libsolidity/lsp/LanguageServer.cpp
Comment (nikola-matic) [inline libsolidity/lsp/LanguageServer.cpp:248]:
```suggestion
```

File: libsolidity/lsp/LanguageServer.cpp
Comment (nikola-matic) [reply libsolidity/lsp/LanguageServer.cpp:237]: By the way, these `get`s can also be replaced with `getOrDefault<string>(jsonPath["prefix"], "")`.

File: libsolidity/lsp/LanguageServer.cpp
Comment (cameel) [reply libsolidity/lsp/LanguageServer.cpp:229]: I'd hold off a bit with using these helpers everywhere. I wanted to get them merged not to stall the asm import PR but in the end we may and up removing them (https://github.com/ethereum/solidity/pull/13579#issuecomment-1286803351). Depends on what we end up with after #13579. I mean, it's fine to use them but for now I'd also not force that.

**Comment by NunoFilipeSantos** [comment]
Right now, we're choosing to close this PR as we won't allocate any time to this particular area.

**Pull request closed without merging**
